Add frdm_mcxe31b board basic support

lucien-nxp opened the pull request:
1. add frdm_mcxe31b board support
2. enable cases below:
    hello_world/blinky/button/philosophers/synchronization
3. add mcxe31x series soc

lucien-nxp committed:
drivers: clock_control: add nxp_mc_cgm clock driver

- add clock_init function to initialize clock sources according
  devicetree settings
- finish basic clock api function

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

lucien-nxp committed:
soc: nxp: mcx: add mcxe31 series soc

- create 'mcxe' as family and 'mcxe31x' as series
- add pinctrl_soc.h
- add soc.c/.h to do some soc level initialization
- add ecc initialization in mcxe31x_soc_initialization.S

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

lucien-nxp committed:
dts: arm: nxp: add mcxe31x device tree

- Generate a full devices device tree file
- Use specific_part.dtsi + full_devices.dtsi way to
  desribe all devices

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

lucien-nxp committed:
drivers: interrupt_controller: adapt for mcxe31x series

- adapt for mcxe31x series
- due to some bit defined in header files
  add some conditional macro to separate

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

lucien-nxp commented:
Hi @hakehuang
Could you help start test for this new board?

hakehuang commented on boards/nxp/frdm_mcxe31B/frdm_mcxe31b.yaml:
```diff
@@ -0,0 +1,19 @@
+  - gnuarmemb
+supported:
+  - gpio
+  - uart
```
please enable tests/drivers/uart_basic_api with below
```
  drivers.uart.async_api.lpuart:
    filter: CONFIG_SERIAL_SUPPORT_ASYNC and CONFIG_UART_MCUX_LPUART and not CONFIG_CPU_HAS_DCACHE
    harness: ztest
    harness_config:
      fixture: gpio_loopback
    depends_on: dma
    extra_configs:
      - CONFIG_USERSPACE=n
      - CONFIG_TEST_USERSPACE=n
    extra_args:
      - platform:frdm_mcxe31b/mcxe31b:"DTC_OVERLAY_FILE=nxp/dut_lpuart<need check schematic>_loopback.overlay"
```

lucien-nxp replied:
Enabled.

hakehuang commented:
@lucien-nxp , there are some cmsis-dsp test failure, due to buffer allocation failure.
```
START - test_arm_cifft_q31_step_2048

    Assertion failed at WEST_TOPDIR/zephyr/tests/lib/cmsis_dsp/transform/src/cq31.c:142: test_arm_cifft_q31: (scaled_ref is NULL)
buffer allocation failed
 FAIL - test_arm_cifft_q31_step_2048 in 0.014 seconds
```

hakehuang commented on boards/nxp/frdm_mcxe31B/frdm_mcxe31b.yaml:
```diff
@@ -0,0 +1,19 @@
+name: NXP FRDM MCXE31B
+type: mcu
+arch: arm
+ram: 512
```
you put dtcm as zephyr,sram, which is only 64K

lucien-nxp replied:
Update to use sram region as default ram, and size is 288K.

manuargue commented on include/zephyr/dt-bindings/clock/nxp_mc_cgm.h:
The content of this file is specific to a single platform, therefore shouldn't be generically called `nxp_mc_cgm.h` because it doesn't apply for all devices featuring the MC_CGM hw IP.

lucien-nxp replied:
If in the future a new npi is equipped with the mc_cgm IP, there is no need to maintain multiple files. Just simply modifying this common file will suffice.

manuargue replied:
All defines in this file are with `MCXE31X`. What would be the way to support multiple platforms? guarding them? Having multiple files looks cleaner to me.

mmahadevan108 replied:
@lucien-nxp , please rename the macro `MCUX_MCXE31X_CLK_ID` to something more generic.

manuargue commented on soc/nxp/mcx/mcxe/mcxe31x/pinctrl_soc.h:
```diff
@@ -0,0 +1,33 @@
+#include <zephyr/dt-bindings/pinctrl/nxp-s32-pinctrl.h>
+#include <zephyr/sys/util.h>
+
+#include "../../../s32/common/siul2_pinctrl.h"
```
I'd suggest moving this file to an NXP common directory. Furthermore, because the SIUl2 driver is now native and doesn't depend on MCUX or S32 SDK's, we should remove the S32 references from all places.
We can defer to another PR to not bloat this one, but I wanted to mention it since this is the first non-S32 platform that is sharing a lot of hw IPs with the S32.

manuargue commented on include/zephyr/drivers/clock_control/nxp_clock_control.h:
ditto for this file, it's specific to an SoC, let's not name it generically "nxp_clock_control.h"

lucien-nxp replied:
No, I want to maintain a common nxp clock.h file here, any devices related to clock can be get parameter here. I refer ST company usage to create one for nxp product. From now on, when we obtain clock parameters, we can directly call this .h file, which makes it easier to obtain certain parameters(The premise is that it is maintained by all of us together.)

manuargue commented on drivers/interrupt_controller/intc_eirq_nxp_s32.c:
```diff
@@ -28,13 +28,20 @@
 #define SIUL2_IFER0             0x28
 /* SIUL2 Interrupt Filter Maximum Counter Register */
 #define SIUL2_IFMCR(n)          (0x30 + 0x4 * (n))
+#ifndef SIUL2_IFMCR_MAXCNT_MASK
```
where are these macros being defined for mcxe31x?
this native driver requires to have all definitions in one place

manuargue replied:
This is a full zephyr native driver, it doesn't depend on hal_nxp headers. Reason being that the NXP HAL's are not consistent in defining these headers. MCUX SDK does it one way, RTD (S32) does it another way. I'd like to keep this decoupling and not re-introduce HAL into the mix.

ZhaoxiangJin replied:
IMO, native driver means without invoking SDK/RTD driver APIs instead of not using the device header. Mask like 'SIUL2_IFMCR_MAXCNT_MASK' is defined in device header not SDK SIUL2 driver header, so i think the current implementation is fine. what is your opinion @decsny?

manuargue replied:
agreed to keep it this way for now and we'll look at it after merge if there's a better way to integrate this

lucien-nxp commented on soc/nxp/mcx/mcxe/mcxe31x/Kconfig.defconfig:
```diff
@@ -0,0 +1,16 @@
+	default 240
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 160000000 if CORTEX_M_SYSTICK
```
yes, this is a good ideal for getting os tick per from dts. However, CPU frequency is determined by specific board setting, I don't know how to get a frequency from dts because nobody will record the cpu frequency in dts. 
@hakehuang  @decsny , I don't find an example in NXP platform for record cpu clock/systick clock, do you have any suggestion?

nordicjm replied:
see https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/nxp/frdm_mcxc444/frdm_mcxc444.dts#L102 and https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/atmel/sam/Kconfig.defconfig#L18

lucien-nxp committed:
driver: pinctrl: adapt for mcxe31x series

- add binding files: nxp,mcxe31x-siul2-pinctrl.yaml
- Enable PINCTRL_NXP_SIUL2 when nxp,mcxe31x-siul2-pinctrl is ok

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

lucien-nxp commented:
Hi @hakehuang 
Could you help do a new full round test for frdm-mcxe31b?  I have added ECC initialization code in this PR.

lucien-nxp commented on soc/nxp/mcx/mcxe/mcxe31x/Kconfig:
```diff
@@ -0,0 +1,47 @@
+	default "ITCM"
+	help
+	  Select memory to relocate sram_config.c code
+endif # CODE_DATA_RELOCATION
```
Updated.

lucien-nxp commented on dts/arm/nxp/nxp_mcxe31b.dtsi:
```diff
@@ -0,0 +1,20 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
```
Updated.

lucien-nxp commented on boards/nxp/frdm_mcxe31B/boot_header/boot_header.c:
```diff
@@ -0,0 +1,112 @@
+extern void *const _vector_start;
+#else
+#error Unsupported toolchain!
+#endif
```
Updated.

lucien-nxp commented on boards/nxp/frdm_mcxe31B/boot_header/boot_header.c:
```diff
@@ -0,0 +1,112 @@
+ * Boot header
+ ******************************************************************************/
+typedef struct image_vector_table {
+	uint32_t Header;                    /* Header                         */
```
Updated.

lucien-nxp commented on west.yml:
```diff
@@ -210,7 +210,7 @@ manifest:
         - hal
     - name: hal_nxp
-      revision: c2279acd1e06aa17926cc8bd89812be691dd005f
+      revision: pull/599/head
```
Updated.

lucien-nxp commented:
> @lucien-nxp , there are some cmsis-dsp test failure, due to buffer allocation failure.
> 
> ```
> START - test_arm_cifft_q31_step_2048
> 
>     Assertion failed at WEST_TOPDIR/zephyr/tests/lib/cmsis_dsp/transform/src/cq31.c:142: test_arm_cifft_q31: (scaled_ref is NULL)
> buffer allocation failed
>  FAIL - test_arm_cifft_q31_step_2048 in 0.014 seconds
> ```

Resolve the issue on the newest PR, please help re-trigger a new round test for frdm_mcxe31b. Thank you.

lucien-nxp commented on boards/nxp/frdm_mcxe31B/boot_header/boot_header.c:
```diff
@@ -0,0 +1,112 @@
+		      CLR(APP_SWT_INIT_MASK),  /* SWT0 is not setup by BAF        */
+	.CM7_0_StartAddress =
+#if defined(__CC_ARM) || defined(__ARMCC_VERSION)
+		(const uint32_t *)&__Vectors,
```
Updated.

lucien-nxp commented on soc/nxp/mcx/mcxe/Kconfig.defconfig:
```diff
@@ -0,0 +1,6 @@
+
+# Source series Kconfig files first, so SOCs
+# can override the defaults given here
+rsource "*/Kconfig.defconfig"
```
Updated.

lucien-nxp committed:
boards: nxp: add frdm_mcxe31b board support

- support XIP way to boot
- add board doc and picture
- enable cases below:
    hello_world/blinky/button/
    philosophers/synchronization/
    gpio_basic_api

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

lucien-nxp commented on soc/nxp/mcx/mcxe/mcxe31x/Kconfig:
```diff
@@ -0,0 +1,49 @@
+	select SOC_EARLY_INIT_HOOK
+	select CODE_DATA_RELOCATION
+
+if CODE_DATA_RELOCATION
```
Updated.

lucien-nxp commented on soc/nxp/mcx/mcxe/mcxe31x/Kconfig:
```diff
@@ -0,0 +1,49 @@
+	  ROM requires a fixed IVT offset for each type of boot device.
+
+config ROM_START_OFFSET
+	default 0x1000
```
Moved.

lucien-nxp commented on boards/nxp/frdm_mcxe31B/boot_header/boot_header.c:
```diff
@@ -0,0 +1,102 @@
+	const uint32_t *reserved5;          /* Reserved                       */
+	const uint32_t *reserved6;          /* Reserved                       */
+	const uint32_t *lcc_config;           /* Address of LC config           */
+	uint8_t Reserved7[216];             /* Reserved for future use        */
```
Sorry, miss it. I have modified and use AI tool help check, there is no camel issue on boot_header files

lucien-nxp commented on boards/nxp/frdm_mcxe31b/frdm_mcxe31b-pinctrl.dtsi:
```diff
@@ -0,0 +1,27 @@
+			pinmux = <PTE14_LPUART5_TX_O>;
+			output-enable;
+		};
+		group2 {
```
Added

lucien-nxp commented on boards/nxp/frdm_mcxe31b/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+#
+# Copyright 2025 NXP
```
Updated

lucien-nxp replied:
Fixed. it seems the display issue?<img width="1037" height="672" alt="image" src="https://github.com/user-attachments/assets/b08d5c4a-925c-4ddc-a632-c50b5df279be" />

lucien-nxp replied:
space in comment title have been remove. Thank you, I misunderstand at first.

kartben commented on boards/nxp/frdm_mcxe31b/doc/frdm_mcxe31b.webp:
ouch, the image quality is pretty terrible :)
How about the below instead (also has transparent background as a bonus 😄)

![Image](https://github.com/user-attachments/assets/259617c5-fdd6-4c6f-bc5b-50ec55976361)

lucien-nxp replied:
Yes, this looks better! Thank you. I have already replaced it with this picture.

hakehuang reviewed (approved):
board testing pass on v4.2.0-5922-gcbc9648808a6

nordicjm commented on dts/arm/nxp/nxp_mcxe31x_common.dtsi:
```diff
@@ -0,0 +1,1063 @@
+			compatible = "soc-nv-flash";
+			reg = <0x400000 DT_SIZE_K(4096)>;
+		};
+
```
```suggestion
```

lucien-nxp replied:
Deleted.

nordicjm commented on soc/nxp/mcx/mcxe/mcxe31x/Kconfig:
```diff
@@ -0,0 +1,50 @@
+	help
+	  The Image Vector Table (IVT) provides the boot ROM with pointers to
+	  the application entry point and device configuration data. The boot
+	  ROM requires a fixed IVT offset for each type of boot device.
```
You are locking others out from using these for other purposes by hogging symbols others might want to use, but since they already exist I will accept it. The board part no, have a Kconfig that controls if these are active in the SoC file and have the board select it, the SoC should not depend on the board and board depend on the SoC, that is a circular dependency, board depends on SoC not the other way around

lucien-nxp commented:
> @lucien-nxp can you address the merge conflicts

Thank you for reminding. Conflict have been resolved.

ZhaoxiangJin reviewed (approved):
Good work, we can start feature enablement on FRDM-MCXE31B.

hakehuang reviewed (approved):
boards regression pass

JarmouniA commented on dts/arm/nxp/nxp_mcxe31x_common.dtsi:
```diff
@@ -0,0 +1,1062 @@
+		status = "disabled";
+	};
+
+	lcu_0: lcu@98000 {
```
!??

lucien-nxp replied:
As above.
